### PR TITLE
Python client for Model-as-a-Service (MaaS) / Model-as-a-Platform (MaaP) endpoints

### PR DESCRIPTION
**Pull request by dargilco**
# Description

New Python SDK for AI model inference using Model-as-a-Service (MaaS) or Model-as-a-Platform (MaaP) endpoints. It defines ChatCompletionsClient, EmbeddingsClient and ImageEmbeddingsClient, as well as a package function load_client(). It's targeting the endpoint routes /chat/completions, /embeddings/ and /images/embeddings.

[Link to REST API that the SDK is using](https://learn.microsoft.com/en-us/azure/ai-studio/reference/reference-model-inference-api?tabs=azure-studio)
[Link to the TypeSpec PR used to auto-gen this client library](https://github.com/Azure/azure-rest-api-specs/pull/28438)

----

# All SDK Contribution checklist:
- [v ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [ v] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/ai/azure-ai-inference/azure/ai/inference/_operations/_operations.py [hunk 1]
```diff
@@ -0,0 +1,1087 @@
...
+                                "role": "str",  # The chat role associated with the
+                                  message. Required. Known values are: "system", "user", "assistant",
+                                  and "tool".
+                                "tool_calls": [
+                                    chat_completions_tool_call
+                                ]
+                            }
+                        }
+                    ],
+                    "created": "2020-02-20 00:00:00",  # The first timestamp associated with
```
Comment (johanste) [inline sdk/ai/azure-ai-inference/azure/ai/inference/_operations/_operations.py:377]: This example is incorrect/inconsistent. The wire format should be an integer. The string format should be a datetime/iso 8601 formatted value. Need to track down where this comes from - but @iscai-msft may beat me to it if I tag her here...

File: sdk/ai/azure-ai-inference/azure/ai/inference/_client.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    :type credential: ~azure.core.credentials.AzureKeyCredential
+    :keyword endpoint: Service host. Required.
+    :paramtype endpoint: str
+    :keyword api_version: The API version to use for this operation. Default value is
+     "2024-04-01-preview". Note that overriding this default value may result in unsupported
+     behavior.
+    :paramtype api_version: str
+    """
+
+    def __init__(self, credential: AzureKeyCredential, *, endpoint: str, **kwargs: Any) -> None:
```
Comment (johanste) [inline sdk/ai/azure-ai-inference/azure/ai/inference/_client.py:35]: We should support more than key credentials, no?

Comment (dargilco) [reply sdk/ai/azure-ai-inference/azure/ai/inference/_client.py:35]: Yes, eventually, but at the moment only key credentials is supported. Entra ID will be supported later, not sure when. I will try to get clarifications on the timeline.

File: sdk/ai/azure-ai-inference/azure/ai/inference/models/_patch.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+            # If you reached here, the line should contain `data: {...}\n`
+            # where the curly braces contain a valid JSON object. Deserialize it into a ChatCompletionsUpdate object
+            # and add it to the queue.
+            self._queue.put(
+                # pylint: disable=W0212 # Access to a protected member _deserialize of a client class
+                _models.ChatCompletionsUpdate._deserialize(json.loads(line[len(self.SSE_DATA_EVENT_PREFIX) : -1]), [])
+            )
+
+            if self.ENABLE_CLASS_LOGS:
```
Comment (johanste) [inline sdk/ai/azure-ai-inference/azure/ai/inference/models/_patch.py:132]: No print statements please. If you want to log things (at a debug level), use [`logging`](https://docs.python.org/3/library/logging.html).

File: sdk/ai/azure-ai-inference/azure/ai/inference/models/_patch.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+    def __init__(self, response: Union[HttpResponse, AsyncHttpResponse]):
+        self._response = response
+        self._bytes_iterator: Union[AsyncIterator[bytes], Iterator[bytes]] = response.iter_bytes()
+        self._is_async_iterator = isinstance(self._response, AsyncHttpResponse)
+        self._queue: "queue.Queue[_models.ChatCompletionsUpdate]" = queue.Queue()
+        self._incomplete_json = ""
+        self._done = False  # Will be set to True when reading 'data: [DONE]' line
+
+    def __iter__(self):
```
Comment (johanste) [inline sdk/ai/azure-ai-inference/azure/ai/inference/models/_patch.py:43]: We want to have separate stream classes. Defining members and raising at runtime defeats typing.

File: sdk/ai/azure-ai-inference/azure/ai/inference/models/_patch.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        if self._queue.empty():
+            raise StopIteration
+        return self._queue.get()
+
+    def _read_next_block(self) -> bool:
+        try:
+            # Use 'cast' to make 'pyright' error go away
+            element = cast(Iterator[bytes], self._bytes_iterator).__next__()
+        except StopIteration:
+            self.close()
```
Comment (johanste) [inline sdk/ai/azure-ai-inference/azure/ai/inference/models/_patch.py:62]: We should make the stream class a context manager. It is dangerous to rely on the client exhausting the stream in order to close the stream... @kristapratico , this is very much related to our discussions around spans/OTEL LLM semantic conventions and streaming.

File: sdk/ai/azure-ai-inference/samples/sample_chat_completions.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        endpoint=endpoint,
+        credential=AzureKeyCredential(key)
+    )
+    # [END create_client]
+
+    # [START chat_completions]
+    # Do a single chat completion operation. This will be a synchronously (blocking) call.
+    result = client.get_chat_completions(
+        messages=[
+            ChatRequestSystemMessage(content="You are an AI assistant that helps people find information."),
```
[the remark this answers is not in the document]
Comment (dargilco) [reply sdk/ai/azure-ai-inference/samples/sample_chat_completions.py]: @johanste I added two new samples to show how to use the create() method with input arguments `body: JSON` and `body: IO[bytes]`. See:
https://github.com/Azure/azure-sdk-for-python/blob/dargilco/azure-ai-inference/sdk/ai/azure-ai-inference/samples/sample_chat_completions_from_input_json.py
https://github.com/Azure/azure-sdk-for-python/blob/dargilco/azure-ai-inference/sdk/ai/azure-ai-inference/samples/sample_chat_completions_from_input_bytes.py

[the remark this answers is not in the document]
Comment (dargilco) [reply sdk/ai/azure-ai-inference/samples/sample_chat_completions.py]: The second sample loads the text file https://github.com/Azure/azure-sdk-for-python/blob/dargilco/azure-ai-inference/sdk/ai/azure-ai-inference/samples/example_chat.json which contains a conversation.

File: sdk/ai/azure-ai-inference/azure/ai/inference/models/_patch.py (hunk 3 above)
Comment (dargilco) [reply sdk/ai/azure-ai-inference/azure/ai/inference/models/_patch.py:137]: @johanste I planned to remove these before PR complete, but instead I switched to using "logger" to log detailed SSE parsing. It's off by default since this is for development debugging. You need to update the code to enable these logs. Thanks!

File: sdk/ai/azure-ai-inference/azure/ai/inference/models/_patch.py (hunk 4 above)
Comment (dargilco) [reply sdk/ai/azure-ai-inference/azure/ai/inference/models/_patch.py]: Agree! I split the result into two classes, once called StreamingChatCompletions and AsyncStreamingChatCompletions.

File: sdk/ai/azure-ai-inference/azure/ai/inference/models/_patch.py (hunk 5 above)
Comment (dargilco) [reply sdk/ai/azure-ai-inference/azure/ai/inference/models/_patch.py]: @johanste @kristapratico The result object (of type StreamingChatCompletions) has an __exit__() and close() methods, which closes the `HttpResponse` object. Is that not enough?

File: sdk/ai/azure-ai-inference/azure/ai/inference/_operations/_operations.py (hunk 1 above)
Comment (dargilco) [reply sdk/ai/azure-ai-inference/azure/ai/inference/_operations/_operations.py:432]: @iscai-msft do you mind having a look at this? Thanks!

Comment (dargilco) [reply sdk/ai/azure-ai-inference/azure/ai/inference/_operations/_operations.py:401]: @johanste @iscai-msft I posted a Teams chat on this (will open a GitHub issue if needed). Let's continue to discuss this issue there. I removed the long JSON description from the ref-doc comment on my hand-written operation method in _patch.py, so this is no longer relevant for the published package.

File: sdk/ai/azure-ai-generative/setup.py [hunk 7]
```diff
@@ -42,7 +42,7 @@
     url="https://github.com/Azure/azure-sdk-for-python",
     keywords="azure, azuresdk, azure sdk",
     classifiers=[
-        "Development Status :: 4 - Beta",
+        "Development Status :: 7 - Inactive",
```
Comment (johanste) [inline sdk/ai/azure-ai-generative/setup.py:45]: Was this intentional? It probably should be a different PR....

File: sdk/ai/azure-ai-inference/azure/ai/inference/aio/_patch.py [hunk 8]
```diff
@@ -0,0 +1,963 @@
...
+     "2024-05-01-preview". Note that overriding this default value may result in unsupported
+     behavior.
+    :paramtype api_version: str
+    :return: ChatCompletionsClient or EmbeddingsClient or ImageEmbeddingsClient
+    :rtype: ~azure.ai.inference.ChatCompletionsClient or ~azure.ai.inference.EmbeddingsClient
+     or ~azure.ai.inference.ImageEmbeddingsClient
+    :raises ~azure.core.exceptions.HttpResponseError
+    """
+
+    client = ChatCompletionsClient(endpoint, credential, **kwargs)  # Pick any of the clients, it does not matter...
```
Comment (johanste) [inline sdk/ai/azure-ai-inference/azure/ai/inference/aio/_patch.py:75]: Use a context manager:

`with ChatCompletionsClient(endpoint, credential, **kwargs) as client:`

File: sdk/ai/azure-ai-inference/azure/ai/inference/aio/_patch.py [hunk 9]
```diff
@@ -0,0 +1,963 @@
...
+    model_info = await client.get_model_info()
+    await client.close()
+
+    _LOGGER.info("model_info=%s", model_info)
+    if model_info.model_type in (None, ""):
```
Comment (johanste) [inline sdk/ai/azure-ai-inference/azure/ai/inference/aio/_patch.py:80]: I don't think we need this, but if we do, then `if not model_info.model_type:` would be more pythonic. Both `None` and an empty string are False:y.

File: sdk/ai/azure-ai-inference/azure/ai/inference/aio/_patch.py [hunk 10]
```diff
@@ -0,0 +1,963 @@
...
+            map_error(status_code=response.status_code, response=response, error_map=error_map)
+            raise HttpResponseError(response=response)
+
+        if _stream:
+            return _models.AsyncStreamingChatCompletions(response)
+
+        return _deserialize(_models.ChatCompletions, response.json())  # pylint: disable=protected-access
+
+    # Cache here the results of get_model_info call
+    _model_info: Optional[_models.ModelInfo] = None
```
Comment (johanste) [inline sdk/ai/azure-ai-inference/azure/ai/inference/aio/_patch.py:465]: This is a class-level attribute. I don't think that is what you intended?

File: sdk/ai/azure-ai-generative/setup.py (hunk 7 above)
Comment (dargilco) [reply sdk/ai/azure-ai-generative/setup.py:45]: Oh, good call, I did not notice it.

File: sdk/ai/azure-ai-inference/azure/ai/inference/_patch.py [hunk 11]
```diff
@@ -0,0 +1,958 @@
...
+        :keyword seed: If specified, the system will make a best effort to sample deterministically
+         such that repeated requests with the
+         same seed and parameters should return the same result. Determinism is not guaranteed.".
+         Default value is None.
+        :paramtype seed: int
+        :return: ChatCompletions for non-streaming, or StreamingChatCompletions for streaming.
+        :rtype: ~azure.ai.inference.models.ChatCompletions or ~azure.ai.inference.models.StreamingChatCompletions
+        :raises ~azure.core.exceptions.HttpResponseError
+        """
+        error_map = {
```
Comment (johanste) [inline sdk/ai/azure-ai-inference/azure/ai/inference/_patch.py:385]: I think we can replace this method body with:

```python
        if body is _Unset:
            body = {
                "frequency_penalty": frequency_penalty,
                "max_tokens": max_tokens,
                "messages": messages,
                "presence_penalty": presence_penalty,
                "response_format": response_format,
                "seed": seed,
                "stop": stop,
                "stream": stream,
                "temperature": temperature,
                "tool_choice": tool_choice,
                "tools": tools,
                "top_p": top_p,
            }
            body.update(model_extras or {})
            return super().complete({k:v for k, v in body.items() if v is not None})
        return super().complete(
            body,
            messages= messages,
            model_extras = model_extras,
            frequency_penalty = frequency_penalty,
            presence_penalty = presence_penalty,
            temperature = temperature,
            top_p = top_p,
            max_tokens = max_tokens,
            response_format = response_format,
            stop = stop,
            stream = stream,
            tools = tools,
            tool_choice = tool_choice,
            seed = seed,
            **kwargs,       
        )
```

(I'm probably missing some mapping)

Comment (johanste) [reply sdk/ai/azure-ai-inference/azure/ai/inference/_patch.py:385]: Yeah, missed the defaulted header for dealing with extra parameters. But still simpled than copying the whole method body, no?

File: sdk/ai/azure-ai-inference/azure/ai/inference/models/_patch.py [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+        if self._ENABLE_CLASS_LOGS:
+            logger.debug("[Reading next block]")
+        try:
+            element = await self._bytes_iterator.__anext__()
+        except StopAsyncIteration:
+            await self.aclose()
+            return True
+        return self._deserialize_and_add_to_queue(element)
+
+    def __enter__(self):
```
Comment (johanste) [inline sdk/ai/azure-ai-inference/azure/ai/inference/models/_patch.py:168]: Why do you need to supper sync context manager?

File: sdk/ai/azure-ai-inference/README.md [hunk 13]
```diff
@@ -0,0 +1,476 @@
...
+
+To update an existing installation of the package, use:
+
+```bash
+pip install --upgrade azure-ai-inference
+```
+
+## Key concepts
+
+### Create and authenticate a client directly, using key
```
Comment (johanste) [inline sdk/ai/azure-ai-inference/README.md:48]: Samples should use entra id. Key based auth is not to be promoted per Microsoft guidelines.

File: sdk/ai/azure-ai-inference/README.md [hunk 14]
```diff
@@ -0,0 +1,476 @@
...
+
+Or for example when you created an `EmbeddingsClient` and called `embed` on the client, but the endpoint does not
+support the `/embeddings` route:
+
+```text
+Status code: 405 (Method Not Allowed)
+Operation returned an invalid status 'Method Not Allowed'
+```
+
+### Logging
```
Comment (johanste) [inline sdk/ai/azure-ai-inference/README.md:413]: I think you can just link to https://learn.microsoft.com/en-us/azure/developer/python/sdk/azure-sdk-logging (but we should see if there isn't an aka.ms link to use)

File: sdk/ai/azure-ai-inference/README.md [hunk 15]
```diff
@@ -0,0 +1,476 @@
...
+pip install --upgrade azure-ai-inference
+```
+
+## Key concepts
+
+### Create and authenticate a client directly, using key
+
+The package includes two clients `ChatCompletionsClient` and `EmbeddingsClient`<!-- and `ImageGenerationClients`-->. Both can be created in the similar manner. For example, assuming `endpoint` and `key` are strings holding your endpoint URL and key, this Python code will create and authenticate a synchronous `ChatCompletionsClient`:
+
+```python
```
Comment (johanste) [inline sdk/ai/azure-ai-inference/README.md:52]: I would suggest starting with a zero-to-hero scenario off the bat. Possibly using

```python
creds = azure.identity.DefaultAzureCredential(exclude_interactive_browser_credential=False)
```

which will pop a browser window if the user isn't already logged in... The goal would be to have something that I can copy/paste and run and not have a bunch of things in between the code blocks. If you need to comment on something, commenting in the code can be done instead of splitting code blocks with markdown blocks.

File: sdk/ai/azure-ai-inference/README.md (hunk 13 above)
Comment (dargilco) [reply sdk/ai/azure-ai-inference/README.md:48]: EntraID is not yet supported on MaaS. Once it's supported, and I make another release, I will switch most samples to use EntraID.

File: sdk/ai/azure-ai-inference/README.md (hunk 15 above)
Comment (dargilco) [reply sdk/ai/azure-ai-inference/README.md:52]: I updated all examples of using DefaultAzureCredential() to DefaultAzureCredential(exclude_interactive_browser_credential=False).

Comment (dargilco) [reply sdk/ai/azure-ai-inference/README.md:52]: Regarding _"The goal would be to have something that I can copy/paste and run and not have a bunch of things in between the code blocks"_ -- I agree! and if you look at the chat completion and text embeddings example sections, you will see fully runnable code there (creating the client and doing chat completions/embeddings). You still need to provide the endpoint and key though.

Comment (dargilco) [reply sdk/ai/azure-ai-inference/README.md:52]: I was thinking about perhaps showing reading endpoint url and key directly from environment variables, in the constructor of the clients. Code in README.me is picked up from the samples. In the samples I do read url and key from environment variables, but it is in a separate "try" block at the start, so I can have a better error message when one of the environment variables is missing. But I don't want that in the code snippets in README.md. Bottom line: I need to think about this some more. For now, as there is urgency in getting a first release out, I prefer to keep the code snippets in  README.md as is. But may change that in the next release based on additional discussion with you and feedback. I hope that's okay with you. Thanks!

File: sdk/ai/azure-ai-inference/README.md (hunk 14 above)
Comment (dargilco) [reply sdk/ai/azure-ai-inference/README.md:413]: At the end of the logging section I have this sentence:
For more information see [Configure logging in the Azure libraries for Python](https://aka.ms/azsdk/python/logging)

Python logging is complicated. I wish there was just some environment variable(s) I can set, instead of adding Python commands to my script. Therefore, I wanted to show the few lines of code that you can just grab and put at the top of your script, and have some text that better describes how to see unredacted logs. I use it myself all the time, and it's easy for me to look at what I wrote in the README.md instead of looking at the more detailed link https://aka.ms/azsdk/python/logging . So I'd rather keep things as is in the README.md with regards to logging.

File: sdk/ai/azure-ai-inference/azure/ai/inference/models/_patch.py (hunk 12 above)
Comment (dargilco) [reply sdk/ai/azure-ai-inference/azure/ai/inference/models/_patch.py:168]: Indeed not needed! I removed the method `__enter__(self)` from the classes `StreamingChatCompletions` and `AsyncStreamingChatCompletions` classes. Thanks!

File: sdk/ai/azure-ai-inference/azure/ai/inference/aio/_patch.py (hunk 10 above)
Comment (dargilco) [reply sdk/ai/azure-ai-inference/azure/ai/inference/aio/_patch.py:465]: Good catch! I was doing C# coding in Python... I removed the statement `_model_info: Optional[_models.ModelInfo] = None`. And updated the methods `get_model_info()` and `__str__()`. As of now, if you create a client directly, it will no have the attribute `_model_info`. Only after calling `get_model_info()`, the attribute `_model_info` will be created and will hold the model info results from the service call. The alternative is to override the `__init__` method in `_patch.py` for all the clients (sync and async clients), and define plus initialize `_model_info` to `None` there. At this point (just to support _model_info), I don't want to override all `__init__` methods.

File: sdk/ai/azure-ai-inference/azure/ai/inference/aio/_patch.py (hunk 9 above)
Comment (dargilco) [reply sdk/ai/azure-ai-inference/azure/ai/inference/aio/_patch.py:80]: Thank you! I replaced with `if not model_info.model_type:`

File: sdk/ai/azure-ai-inference/azure/ai/inference/aio/_patch.py (hunk 8 above)
Comment (dargilco) [reply sdk/ai/azure-ai-inference/azure/ai/inference/aio/_patch.py:75]: Thanks! I now use `async with` here and in all async samples, whenever I create the client. I does not look like I need to use the `with` pattern when creating the sync client in samples, so I left the sync samples as-is for simplicity (I never did show a call to client.close() in the sync samples).

File: sdk/ai/azure-ai-inference/azure/ai/inference/_patch.py (hunk 11 above)
Comment (dargilco) [reply sdk/ai/azure-ai-inference/azure/ai/inference/_patch.py:385]: Unfortunately, we can't do that, as I need special code to deal with the streaming case. For streaming, the return object should be `_models.StreamingChatCompletions(response)` which is not provided by `super().complete()`.

File: sdk/ai/azure-ai-inference/azure/ai/inference/aio/_client.py [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
+            "endpoint": self._serialize.url("self._config.endpoint", self._config.endpoint, "str", skip_quote=True),
+        }
+
+        request_copy.url = self._client.format_url(request_copy.url, **path_format_arguments)
+        return self._client.send_request(request_copy, stream=stream, **kwargs)  # type: ignore
+
+    async def close(self) -> None:
+        await self._client.close()
+
+    async def __aenter__(self) -> "ChatCompletionsClient":
```
Comment (johanste) [inline sdk/ai/azure-ai-inference/azure/ai/inference/aio/_client.py:108]: this should be typed as returning `typing.Self`. Opened https://github.com/Azure/autorest.python/issues/2619 to track that issue. We can override __aenter__ in the patched client(s). This should fix typing when used in context managers.

File: sdk/ai/azure-ai-inference/azure/ai/inference/_patch.py [hunk 17]
```diff
@@ -0,0 +1,952 @@
...
+        content_type: str = "application/json",
+        model_extras: Optional[Dict[str, Any]] = None,
+        frequency_penalty: Optional[float] = None,
+        presence_penalty: Optional[float] = None,
+        temperature: Optional[float] = None,
+        top_p: Optional[float] = None,
+        max_tokens: Optional[int] = None,
+        response_format: Optional[Union[str, _models.ChatCompletionsResponseFormat]] = None,
+        stop: Optional[List[str]] = None,
+        stream: Optional[bool] = None,
```
Comment (johanste) [inline sdk/ai/azure-ai-inference/azure/ai/inference/_patch.py:131]: We should overload on literal `True`/`False` values. That way we can correctly type the return value and avoid the Union.

File: sdk/ai/azure-ai-inference/azure/ai/inference/aio/_client.py (hunk 16 above)
Comment (dargilco) [reply sdk/ai/azure-ai-inference/azure/ai/inference/aio/_client.py:108]: Good job finding the issue here! Thank you. Fixed with commit https://github.com/Azure/azure-sdk-for-python/pull/34937/commits/0883400feacdd2dbca13c208a45b0bda41b63b8b

File: sdk/ai/azure-ai-inference/azure/ai/inference/_patch.py (hunk 17 above)
Comment (dargilco) [reply sdk/ai/azure-ai-inference/azure/ai/inference/_patch.py:192]: Per IM conversation, we stay with one 'complete' method for both streaming and non-streaming. And fixed mypy and pyright errors by adding overloads with `stream: Literal`. Commit: https://github.com/Azure/azure-sdk-for-python/pull/34937/commits/0d03ad2ed5c266deed840ec69a2c2396ef47f5b4

File: sdk/ai/azure-ai-inference/azure/ai/inference/_patch.py [hunk 18]
```diff
@@ -0,0 +1,1037 @@
...
+    @distributed_trace
+    def get_model_info(self, **kwargs: Any) -> _models.ModelInfo:
+        # pylint: disable=line-too-long
+        """Returns information about the AI model.
+
+        :return: ModelInfo. The ModelInfo is compatible with MutableMapping
+        :rtype: ~azure.ai.inference.models.ModelInfo
+        :raises ~azure.core.exceptions.HttpResponseError
+        """
+        if not hasattr(self, "_model_info"):
```
Comment (johanste) [inline sdk/ai/azure-ai-inference/azure/ai/inference/_patch.py:530]: The canonical way to do this is to initialize self._model_info as None in the __init__.

Comment (johanste) [reply sdk/ai/azure-ai-inference/azure/ai/inference/_patch.py:530]: (or, if you want to be fancy, https://docs.python.org/3/library/functools.html#functools.cached_property)

Comment (dargilco) [reply sdk/ai/azure-ai-inference/azure/ai/inference/_patch.py:530]: Yes, but that would require me to override the __init__ method of all the clients, sync and async flavors. That's a bit too risky for me at this late stage. We can consider this again in the next release.

Comment (dargilco) [reply sdk/ai/azure-ai-inference/azure/ai/inference/_patch.py:530]: `cached_property` is neat though!

File: eng/pipelines/templates/stages/platform-matrix-ai.json.old
Comment (johanste) [inline eng/pipelines/templates/stages/platform-matrix-ai.json.old:1]: Was this intentional?

File: sdk/ai/azure-ai-inference/azure/ai/inference/_patch.py (hunk 18 above)
Comment (dargilco) [reply sdk/ai/azure-ai-inference/azure/ai/inference/_patch.py]: Changed my mind! fixed with commit https://github.com/Azure/azure-sdk-for-python/pull/34937/commits/864796470f52b50be43883916f5b1360ca521612

File: eng/pipelines/templates/stages/platform-matrix-ai.json.old
Comment (dargilco) [reply eng/pipelines/templates/stages/platform-matrix-ai.json.old:1]: Yes! but I think I can delete it now.

Comment (dargilco) [reply eng/pipelines/templates/stages/platform-matrix-ai.json.old:1]: I deleted the file.

**Review by johanste — APPROVED** [review]
